src/main/scala/io/hydrosphere/mist/master/MasterServer.scala

```diff
@@ -0,0 +1,233 @@
+    masterService.recoverJobs()
+
+    httpBinding = Some(bootstrapHttp(streamer, masterService))
+
+    // Start MQTT subscriber
+    mqttInput = bootstrapMqtt(masterService)
+
+    // Start Kafka subscriber
+    kafkaInput = bootstrapKafka(masterService)
+  }
+
+  private def bootstrapKafka(masterService: MasterService): Option[AsyncInterface] = {
+    if (config.kafka.isOn) {
+      import config.kafka._
+
+      val input = AsyncInput.forKafka(host, port, subscribeTopic)
+      val interface = new AsyncInterface(masterService, input, Source.Async("Kafka")).start()
+      logger.info("Kafka interface is started")
+      Some(interface)
+    } else {
+      None
+    }
+  }
+
+  private def bootstrapMqtt(masterService: MasterService): Option[AsyncInterface] = {
+    if (config.mqtt.isOn) {
+      import config.mqtt._
+
+      val input = AsyncInput.forMqtt(host, port, subscribeTopic)
+      val interface = new AsyncInterface(masterService, input, Source.Async("Mqtt")).start()
+      logger.info("Mqtt interface is started")
+      Some(interface)
+    } else {
+      None
+    }
+  }
+
+  private def bootstrapHttp(streamer: EventsStreamer, masterService: MasterService): ServerBinding = {
+    val http = {
+      val api = new HttpApi(masterService)
+      val apiv2 = {
+        val api = HttpV2Routes.apiWithCORS(masterService)
+        val ws = new WSApi(streamer)
+        // order is important!
+        // api router can't chain unhandled calls, because it's wrapped in cors directive
+        ws.route ~ api
+      }
+      val http = new HttpUi(config.http.uiPath).route ~ api.route ~ DevApi.devRoutes(masterService) ~ apiv2
+      Http().bindAndHandle(http, config.http.host, config.http.port)
+    }
+    logger.info(s"REST started on ${config.http.host}:${config.http.port}")
+    Await.result(http, Duration.Inf)
+  }
+
+  private def bootstrapSecurity: Option[KInitLauncher.LoopedProcess] = if (config.security.enabled) {
+    import config.security._
+
+    logger.info("Security is enabled - starting Knit loop")
+    val ps = KInitLauncher.create(keytab, principal, interval)
+    ps.run().onFailure({
+      case e: Throwable =>
+        logger.error("KInit process failed", e)
+        stopUnexpectedly()
+    })
+    logger.info("Knit loop started")
+    Some(ps)
+  } else {
+    None
+  }
+
+  def stopUnexpectedly(): Unit = {
+    logger.info("Stop unexpectedly")
+    sys.exit(1)
+  }
+
+  def stop(): Future[Unit] = {
+    logger.info("Stopping mist master")
+    for {
+      _ <- {
+        optionSwitch(security.map(_.stop()))
+      }
+      _ <- {
+        logger.info("Unbinding http port")
+        optionSwitch(httpBinding.map(_.unbind()))
+      }
+      _ <- {
+        logger.info("Unbinding log service")
+        optionSwitch(logsService.map(_.close()))
+      }
+      _ = workerManager.foreach(_ ! StopAllWorkers)
+      _ = eventPublishers.foreach(_.close())
+      _ = mqttInput.foreach(_.close())
+      _ = kafkaInput.foreach(_.close())
```

Comment:
At first, we should stop all interfaces (kafka, mqtt, http), then workers, and then log service.